## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# View the column names in the data


In [4]:
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use the describe function to gather some basic statistics


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [7]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [8]:
df['category'].value_counts()

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [9]:
df['category'].value_counts().nlargest(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [10]:
# What three item categories had the most entries?


In [11]:
df["subcategory"].value_counts().nlargest(10)

subcategory
bathroom supplies     6424
paper products        5874
kitchen supplies      5746
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              3093
Name: count, dtype: int64

In [12]:
most_common_subcategory = df.loc[df["category"] == "consumables" , ["subcategory"]]
most_common_subcategory.value_counts().nlargest(1)

subcategory      
bathroom supplies    6424
Name: count, dtype: int64

In [13]:
# For the category with the most entries, which subcategory had the most entries?


In [14]:
df['client_id'].value_counts().nlargest(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [15]:
# Which five clients had the most entries in the data?


In [16]:
top_5 = df['client_id'].value_counts().nlargest(5).index.tolist()
top_5

[33615, 66037, 46820, 38378, 24741]

In [17]:
# Store the client ids of those top 5 clients in a list.


In [18]:
df.loc[df['client_id'] == 33615, 'qty'].sum()

64313

In [19]:
# How many total units (the qty column) did the client with the most entries order order?


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [20]:
df['line_subtotal'] = df['unit_price'] * df['qty']
df[['unit_price', 'qty', 'line_subtotal']].head(2)


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [21]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty


In [22]:
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df.apply(lambda row: 7 * row['total_weight'] if row['total_weight'] > 50 else 10 * row['total_weight'], axis=1)
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [23]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.



In [24]:
def cal_line_price(row):
    cost = (row['line_subtotal']+row['shipping_price'])
    tax = cost * .0925
    total = round(cost+tax, 2)
    return total

    
    
df['line_price'] = df.apply(cal_line_price, axis=1)
df[['line_subtotal', 'shipping_price', 'line_price']].head(3)

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [25]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%


In [26]:
df['line_price'] = df.apply(cal_line_price, axis=1)
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [27]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).



In [28]:
df['line_profit'] = df['line_price'] - df['line_cost']
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

In [29]:
# Create a column for the profit of each line using line cost and line price



## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [30]:
order1 = df.loc[df['order_id'] == 2742071, 'line_price'].sum().round(2)
order2 = df.loc[df['order_id'] == 2173913, 'line_price'].sum()
order3 = df.loc[df['order_id'] == 6128929, 'line_price'].sum()

print('Order 2742071 total: $', order1)
print('Order 2173913 total: $', order2)
print('Order 6128929 total: $', order3)


Order 2742071 total: $ 152811.89
Order 2173913 total: $ 162388.71
Order 6128929 total: $ 923441.25


In [31]:
# Check your work using the totals above


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [32]:
# df['client_id'].value_counts().nlargest(5)
# top_5
# df.loc[df['client_id'] == 33615, 'line_price'].sum()


for id in top_5:
    summary = df.loc[df['client_id'] == id, 'line_price'].sum().round(2)
    
    print(f' {id} : ${summary}') 
    
    
    
    

 33615 : $8377308.52
 66037 : $10259514.8
 46820 : $9743794.33
 38378 : $12906550.87
 24741 : $82268892.04


In [33]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [34]:

# top_5
# col = ['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']
# client_df = []
# for id in top_5:
#     totals = df.loc[df['client_id'] == id, col].sum()

#     client_df.append(totals)
    
#     # totals_list = totals.tolist(totals)

# client_df

summary_data = []

for client_id in top_5:
    totals = df.loc[df['client_id'] == client_id, ['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']].sum()
    totals['client_id'] = client_id
    summary_data.append(totals)

summary_df = pd.DataFrame(summary_data)
summary_df = summary_df[['client_id', 'qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']]
summary_df.round(2)
# top5_df = pd.DataFrame(totals, columns = col)


# sum_df = df[['client_id', 'qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']]


# sum_df.head(5)


client_id       qty  shipping_price   line_price    line_cost  line_profit
0    33615.0   64313.0      1828984.89   8377308.52   6175313.91   2201994.61
1    66037.0   43018.0      1395151.85  10259514.80   7004482.98   3255031.82
2    46820.0   75768.0      1601448.84   9743794.33   7007191.64   2736602.69
3    38378.0   73667.0      3429455.40  12906550.87   9634720.98   3271829.89
4    24741.0  239862.0      5126448.37  82268892.04  45688899.71  36579992.33

In [35]:
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'line_subtotal', 'total_weight', 'shipping_price',
       'line_price', 'line_cost', 'line_profit'],
      dtype='object')

In [36]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [37]:
renamed_df = summary_df.rename(columns = {'client_id': 'Client ID' , 'qty': 'Units', 'shipping_price': 'Shipping (millions)',\
                             'line_price': 'Total Revenue (millions)', 'line_cost': 'Total Cost (millions)',\
                             'line_profit': 'Total Profit (millions)'})

def scale_to_millions(row):
    return row / 1000000



money_columns = ['Shipping (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']
renamed_df[money_columns] = renamed_df[money_columns].apply(scale_to_millions, axis=1)
renamed_df

Client ID     Units  Shipping (millions)  Total Revenue (millions)  \
0    33615.0   64313.0             1.828985                  8.377309   
1    66037.0   43018.0             1.395152                 10.259515   
2    46820.0   75768.0             1.601449                  9.743794   
3    38378.0   73667.0             3.429455                 12.906551   
4    24741.0  239862.0             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [38]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# # Define a function that converts a dollar amount to millions.


# # Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [43]:
sorted_df = renamed_df.sort_values('Total Profit (millions)', ascending = False)
sorted_df

Client ID     Units  Shipping (millions)  Total Revenue (millions)  \
4    24741.0  239862.0             5.126448                 82.268892   
3    38378.0   73667.0             3.429455                 12.906551   
1    66037.0   43018.0             1.395152                 10.259515   
2    46820.0   75768.0             1.601449                  9.743794   
0    33615.0   64313.0             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

In [40]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


In [41]:
#The total profit is much higher on client id 24741 due to the large amount of units. Tax on large amounts of units sold is what raises a large amount of revenue or the total cost/profit by a large amount.